### PR TITLE
[ADD] Ability to track real user to Environment

### PR DESCRIPTION
**Pull request by parisiale**
Description of the issue/feature this PR addresses:

With this patch we add an optional arg to
the constructor of the Environment class in
in order to keep track of the real user that
is logged in, even after a sudo call.

The logic for such functionality is implemented
in the caller method (as it's invoked by the
BaseModel::sudo method) and is transparent for
the caller of sudo().

Ticket: 3376
User-story: 3376

Current behavior before PR:

Invoking `with_context` after sudo may incur in cache
errors.

Desired behavior after PR is merged:

The real user is stored in the environment
even after a sudo call. No need for using
`with_context`.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by crwr45 — APPROVED** [review]
Looks good to me.

File: odoo/api.py [hunk 1]
```diff
@@ -726,19 +726,20 @@ def reset(cls):
...
         env, envs = None, cls.envs
         for env in envs:
-            if env.args == args:
+            if env.args == args and env.real_uid == real_uid:
                 return env
 
         # otherwise create environment, and add it in the set
         self = object.__new__(cls)
         self.cr, self.uid, self.context = self.args = (cr, uid, frozendict(context))
+        self.real_uid = real_uid
```
Comment (spenced) [inline odoo/api.py:743]: If Env() is called directly then self.real_uid will be None. Is that what you intended?
Should it be set to self.uid instead? That would make its value always an integer id. (If so, real_uid would have to be evaluated before using it in the above condition.)

File: odoo/api.py [hunk 2]
```diff
@@ -786,7 +787,11 @@ def __call__(self, cr=None, user=None, context=None):
         cr = self.cr if cr is None else cr
         uid = self.uid if user is None else int(user)
         context = self.context if context is None else context
-        return Environment(cr, uid, context)
+        return Environment(cr, uid, context, self.get_real_uid())
+
+    def get_real_uid(self):
```
Comment (spenced) [inline odoo/api.py]: If self.real_uid is changed to always be an integer id, then this function can be removed or restricted to a property. (The value of real_uid is constant.)

File: odoo/api.py (hunk 1 above)
Comment (parisiale) [reply odoo/api.py:743]: The original intention was to not expose the attr and rely on the getter for the update.
Initialising it to the current uid would make the getter superfluous as you mentioned, but we'll need to move the update check to the caller method. No strong opinion.

**Comment by parisiale** [comment]
Updated and tested.

File: odoo/api.py [hunk 3]
```diff
@@ -726,19 +726,20 @@ def reset(cls):
...
-    def __new__(cls, cr, uid, context):
+    def __new__(cls, cr, uid, context, real_uid=None):
         assert context is not None
         args = (cr, uid, context)
 
         # if env already exists, return it
         env, envs = None, cls.envs
         for env in envs:
-            if env.args == args:
+            if env.args == args and env.real_uid == real_uid:
```
Comment (spenced) [inline odoo/api.py:737]: `env.real_uid` is always an integer. The default value of `real_uid` is None, which will never match an existing environment.
The calculation of real uid on line 742 should be moved up to line 732 (before the test), e.g.

`real_uid = uid if real_uid is None else real_uid`

(Attribute assignment into self should remain at 742.)

**Review by spenced — APPROVED** [review]
lgtm

**Pull request merged**